pkg/clockcache/cache.go

```diff
@@ -0,0 +1,201 @@
+
+	for idx := range keys {
+		keys[idx], values[idx], _ = self.insert(keys[idx], values[idx])
+	}
+	return
+}
+
+func (self *Cache) insert(key interface{}, value interface{}) (canonicalKey interface{}, canonicalValue interface{}, inserted bool) {
+	elem, present := self.elements[key]
+	if present {
+		return elem.key, elem.Value, false
+	}
+
+	var insertLocation *Element
+	if len(self.storage) >= cap(self.storage) {
+		insertLocation = self.evict()
+		*insertLocation = Element{key: key, Value: value}
+	} else {
+		self.storage = append(self.storage, Element{key: key, Value: value})
+		insertLocation = &self.storage[len(self.storage)-1]
+	}
+
+	self.elements[key] = insertLocation
+	return key, value, true
+}
+
+func (self *Cache) evict() (insertPtr *Element) {
+	for {
+		insertLocation, evicted := self.tryEvict()
+		if evicted {
+			return insertLocation
+		}
+	}
+}
+
+func (self *Cache) tryEvict() (insertPtr *Element, evicted bool) {
+	if self.next >= len(self.storage) {
+		self.next = 0
+	}
+
+	evicted = false
+	reachedEnd := false
+	for !evicted && !reachedEnd {
+		elem := &self.storage[self.next]
+		if elem.used != 0 {
+			elem.used = 0
+		} else {
+			insertPtr = elem
+			key := elem.key
+			delete(self.elements, key)
+			evicted = true
+		}
+		self.next += 1
+		reachedEnd = self.next >= len(self.storage)
+	}
+
+	return
+}
+
+// tries to get a batch of keys and store the corresponding values is valuesOut
+// returns the number of keys that were actually found.
+// NOTE: this function does _not_ preserve the order of keys; the first numFound
+//       keys will be the keys whose values are present, while the remainder
+//       will be the keys not present in the cache
+func (self *Cache) GetValues(keys []interface{}, valuesOut []interface{}) (numFound int) {
+	if len(keys) != len(valuesOut) {
+		panic(fmt.Sprintf("keys and values are not the same len. %d keys, %d values", len(keys), len(valuesOut)))
+	}
+	valuesOut = valuesOut[:len(keys)]
+	n := len(keys)
+	idx := 0
+
+	self.lock.RLock()
+	defer self.lock.RUnlock()
+
+	for idx < n {
+		value, found := self.get(keys[idx])
+		if !found {
+			if n == 0 {
+				return 0
+			}
+			// no value found for key, swap the key with the last element, and shrink n
+			n -= 1
+			keys[n], keys[idx] = keys[idx], keys[n]
+			continue
+		}
+		valuesOut[idx] = value
+		idx += 1
+	}
+	return n
+}
+
+func (self *Cache) Get(key interface{}) (interface{}, bool) {
+	self.lock.RLock()
+	defer self.lock.RUnlock()
+	return self.get(key)
+}
+
+func (self *Cache) get(key interface{}) (interface{}, bool) {
+
+	elem, present := self.elements[key]
+	if !present {
+		return 0, false
+	}
+
+	if atomic.LoadUint32(&elem.used) == 0 {
+		atomic.StoreUint32(&elem.used, 1)
+	}
+
+	return elem.Value, true
+}
+
+func (self *Cache) Unmark(key string) bool {
+	self.lock.RLock()
+	defer self.lock.RUnlock()
+
+	elem, present := self.elements[key]
+	if !present {
+		return false
+	}
+
+	if atomic.LoadUint32(&elem.used) != 0 {
```

Comment:
if you're talking about the entire `if` then see https://github.com/timescale/timescale-prometheus/pull/147/files#r453666623.
If you're worried about races, then there are two cases:

1. A `Get` sets used to `1` after we perform the `Load`. This doesn't matter, since the value remains the same
2. Another `Unmark` sets the value to `0` after the `Load`. Both threads write the same value, so it's idempotent.

Notably the evictor cannot be running because of the lock.